File: 7-grafos/3356.py
troca={}
[x,y,z]=list(map(int,input().split()))
adj=[[] for i in range(x)]
for i in range(y):
    v=input().split()
    for j in range(3):
        troca[v[j]]=len(troca) if v[j] not in troca else troca[v[j]]
    adj[troca[v[2]]][0:0]=troca[v[0]],troca[v[1]]
for i in range(z):
    [v1,v2]=input().split()
    andados=[0 for i in range(x)]
    andados[troca[v1]]=andados[troca[v2]]=1
    fila=[troca[v1],troca[v2]]
    sim=False
    while True:
        for j in adj[fila[-1]]:
            if andados[j]==1:
                sim=True
                fila.clear()
                break
            fila.insert(0,j)
            andados[j]=1
        if sim:break
        fila.pop()
        if len(fila)==0:
            break
    print("verdadeiro") if sim else print("falso")

File: 7-grafos/3356.cpp
#include<iostream>
#include<vector>
#include<list>
#include<map>
#include<cstdlib>
//
int main(){
    std::vector<std::vector<int>>l_adj;
    std::map<std::string,int>troca;
    std::string inp[3];
    int v[3],tam=0;
    std::cin>>v[0]>>v[1]>>v[2];
    l_adj.resize(v[0]);
    for(int i=0;i<v[1];++i){
        for(int j=0;j<3;++j){
            std::cin>>inp[j];
            if(troca.find(inp[j])==troca.end()){
                troca[inp[j]]=tam;
                ++tam;
            }
        }
        l_adj[troca[inp[2]]].insert(l_adj[troca[inp[2]]].end(),{troca[inp[0]],troca[inp[1]]});
    }
    bool sim=false;
    int*andados;
    for(int i=0;i<v[2];++i){
        sim=false;
        andados=(int*)calloc(v[0],sizeof(int));
        std::cin>>inp[0]>>inp[1];
        //
        std::cin.clear();
        std::cin.ignore(2147483647,'\n');
        //
        std::list<int>fila{troca[inp[0]],troca[inp[1]]},aux;
        andados[troca[inp[0]]]=andados[troca[inp[1]]]=1;
        for(;;){
            for(int bgl:l_adj[fila.back()]){
                if(andados[bgl]==1){
                    sim=true;
                    aux.clear();
                    fila.clear();
                    break;
                }
                andados[bgl]=1;
                aux.push_back(bgl);
            }
            if(fila.size()==0&&aux.size()==0)
                break;
            fila.pop_back();
            if(fila.size()==0){
                fila.merge(aux);
                aux.clear();
            }
        }
        if(sim)
            std::cout<<"verdadeiro\n";
        else
            std::cout<<"falso\n";
        delete andados;
        andados=NULL;
    }
}